### ISSUE TITLE
nn: revisit define-layer syntax for attention and residual client code

### ISSUE THREAD

**Issue by bkc39** [bkc39/rktorch#117]
## Problem

The `define-layer` syntax from #116 gives us a useful registered layer tree,
but real client code for residual networks and attention still has too much
ceremony. Each stored child must be named in the field header and then
initialized again in `#:init`:

```racket
(define-layer TransformerBlock (norm1 attention norm2 mlp)
  #:init (width heads max-t #:dropout [dropout 0.1])
  (set! norm1 (LayerNorm width))
  (set! attention (CausalSelfAttention width heads max-t #:dropout dropout))
  (set! norm2 (LayerNorm width))
  (set! mlp (FeedForward width #:dropout dropout))
  #:forward (x)
  ...)
```

The duplication becomes more noticeable in multi-head attention, residual
stages, and nested transformer stacks. Derived configuration such as
`head-dim` also has to be declared as a field when it is only needed by the
forward body.

## Proposed direction

Investigate a single-site declaration form that keeps the registered object
tree and parameter naming behavior, while allowing ordinary constructor locals
and helpers to remain lexical:

```racket
(define-layer (TransformerBlock width heads max-t #:dropout [dropout 0.1])
  (define-field norm1 (LayerNorm width))
  (define-field attention
    (CausalSelfAttention width heads max-t #:dropout dropout))
  (define-field norm2 (LayerNorm width))
  (define-field mlp (FeedForward width #:dropout dropout))

  #:forward (x)
  (define h (+ x (attention (norm1 x))))
  (+ h (mlp (norm2 h))))
```

`define-field` is a design sketch, not an API commitment. The implementation
should decide whether this or an explicit `#:state` clause gives the clearest
syntax. Ordinary `define` should remain a local binding and should not be
registered as a parameter, buffer, or child.

## Semantics to settle

- Preserve `Parameter`, `Buffer`, `layer?`, and `LayerList` classification.
- Preserve depth-first parameter ordering and stable checkpoint names.
- Make constructor locals available to `#:forward` without registering them.
- Define whether a field may be absent and how that is represented.
- Define persistent mutable state separately from temporary `set!` bindings in
  forward. KV caches and running normalization statistics need an explicit
  policy.
- Decide how keyword and optional forward arguments should work for
  cross-attention and masking.
- Ensure a closure that captures a child does not silently hide that child from
  parameter discovery, or document the required registration form.
- Preserve constructor contracts and derived lowercase predicates.

## Acceptance examples

Use client code modeled on the following as the design test:

- a ResNet-style residual block with both identity and 1x1 projection
  shortcuts;
- repeated residual stages built with `LayerList`;
- causal self-attention with Q/K/V projections, a registered causal mask,
  dropout, and head reshaping;
- a pre-norm transformer block and stack;
- parameter discovery, checkpoint paths, `train!`/`eval!`, backward, and Adam
  from the resulting layer tree.

The side-by-side Racket, Python, and Jane Street OCaml examples used for the
initial review are in `/tmp/rktorch-layer-review` locally. The relevant design
comparison is in `README.md` and `syntax-options.rkt.txt` there.

## Scope

This is a syntax and semantics follow-up to #116. It should be evaluated with
real attention and residual client code before changing the macro. It does not
require introducing a general `nn.Module` compatibility layer or changing the
native tensor API.

**Comment by bkc39** [comment]
Sequencing note from the #116 merge: the owner wants #119 (example refactor into blocks) first, then #27 (mutable state, so `Dropout` lands on `define-layer`), to get user-space code close to beta-ready before any further syntax change. `define-field` is deliberately deferred until after those two, to avoid changing the syntax twice in quick succession.
